### ISSUE TITLE
Assume `if TYPE_CHECKING: ... else: ...` block is covered

### ISSUE THREAD

**Issue by retnikt** [coveragepy/coveragepy#831]
`if typing.TYPE_CHECKING` blocks are never run but are still counted as needing test coverage.
( https://docs.python.org/3.7/library/typing.html#typing.TYPE_CHECKING )

**Example**
```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    # coverage.py thinks this is not covered because TYPE_CHECKING is True at runtime
    ...
else:
    ....
```

You have to add `# pragma: no_cover` to the `if` block here because coverage.py assumes TYPE_CHECKING to be False, and therefore assumes one of the blocks does not have a test case. It would be good if you didn't have to, and blocks requiring TYPE_CHECKING to be True are assumed to not need test cases.

**Comment by nedbat** [comment]
You can solve this yourself today by adding this to your .coveragerc file:
```
[report]
exclude_lines = 
    pragma: no cover
    if TYPE_CHECKING:
```
More details about this setting are here: [Advanced exclusion](https://coverage.readthedocs.io/en/v4.5.x/excluding.html#advanced-exclusion).

Since this is something that is configurable by the user, I would rather not change coverage.py to deal with it.

**Comment by cjw296** [comment]
@nedbat - when would you expect code inside `if TYPE_CHECKING` blocks to be executed while code coverage measurement is in effect? How would you feel about making this part of the default `exclude_lines`?

**Comment by nedbat** [comment]
A few years later, I think you are right that it should be part of the default.  If the user wants to measure those lines, they can.

**Comment by ikonst** [comment]
Can we also add `...` to the defaults while we're at it?

**Comment by cjw296** [comment]
@ikonst - I'm not 100% sure that's a good idea, but also as I've painfully discovered, you really mean `\.\.\.`.

**Comment by ikonst** [comment]
Yeah, obviously in its escaped form. I think the sharp change in coverage would tell you :)

**Comment by cjw296** [comment]
It can be quite the footgun when your projects are starting from 100% line coverage...

**Comment by Kludex** [comment]
> A few years later, I think you are right that it should be part of the default. If the user wants to measure those lines, they can.

@nedbat Is the `else` from a `if not TYPE_CHECKING:` supposed to be supported as well? If yes, any tips where I can implement this part?

**Comment by ssbarnea** [comment]
While coveragepy itself seems to recognized the excluded lines as covered, it seems that uploading the reports to codecov.io does show them as not covered in their reporting. If someone spotted a workaround for that, please let me know.

**Comment by ikonst** [comment]
I think "excluded" is its own status in coveragepy, are you sure it's "covered" and not "excluded"? Or are you just concerned about coverage %?

**Comment by benjaoming** [comment]
Am I correct that solving this happens in these lines:

https://github.com/nedbat/coveragepy/blob/fcdce1deb2f182dc5a34fdc6d91b50033ed2badb/coverage/config.py#L152-L155

...+ adding test case(s)?

**Comment by nedbat** [comment]
Yes, that is where the default exclusions are defined.

**Comment by nedbat** [comment]
In commit d29f0750e75a5fa5d317f67aab7b4ad443ada72a, I added `if TYPE_CHECKING:` and `...` (don't worry, with the correct regex) as default exclusions.

**Comment by KSmanis** [comment]
@nedbat the regex currently won't cover inline ellipses, e.g.:
```python
class Example(Protocol):
    def __call__(self): ...

@overload
def overloaded_method(): ...
```

**Comment by nedbat** [comment]
> [@nedbat](https://github.com/nedbat) the regex currently won't cover inline ellipses, e.g.:
> 
> class Example(Protocol):
>     def __call__(self): ...
> 
> @overload
> def overloaded_method(): ...

That's true.  I'm wondering if there would be any false positives if we changed it from:
```
    r"^\s*\.\.\.\s*(#|$)",
```
to
```
    r"\s*\.\.\.\s*(#|$)",
```
That would match your lines, but would also match strange things like:
```
print("well, ... #2 is broken")
```
Do you have an idea of a safe regex that would match your inline ellipses?

**Comment by KSmanis** [comment]
I guess one option is `r"(^|\):)\s*\.\.\.\s*(#|$)"`; it's specific enough that false positives would be kept to a minimum.

But probably an easier and better alternative is to extend the list with the `r"class .*\bProtocol\):"` and `r"@(typing\.)?overload"` culprits themselves (as seen at https://coverage.readthedocs.io/en/latest/excluding.html#advanced-exclusion)

**Comment by nedbat** [comment]
> But probably an easier and better alternative is to extend the list with the `r"class .*\bProtocol\):"` and `r"@(typing\.)?overload"` culprits themselves (as seen at https://coverage.readthedocs.io/en/latest/excluding.html#advanced-exclusion)

You are probably right.  I see that in the docs I suggested `class .*\bProtocol\):`, but I actually use `class .*\(Protocol\):`.  Do people actually use Protocol in multiple inheritance?

**Comment by nedbat** [comment]
I guess I'm also a bit concerned that someone could have their own user-defined base class called Protocol, and I'd inadvertently exclude entire classes.

**Comment by KSmanis** [comment]
> Do people actually use Protocol in multiple inheritance?

Random example in the wild: https://mypy.readthedocs.io/en/stable/protocols.html#defining-subprotocols-and-subclassing-protocols

Can be addressed by `r"class .+\(.*\bProtocol\b.*\):"` (without taking into account multi-line cases).

> I guess I'm also a bit concerned that someone could have their own user-defined base class called Protocol, and I'd inadvertently exclude entire classes.

Yeah, regexes can only get you so far; a proper solution would require AST parsing. Maybe just mention it in the docs so that users can apply it on their own?

**Comment by nedbat** [comment]
I just learned that ruff will reformat code so that this:
```
def foo():
    ...
```
becomes:
```
def foo(): ...
```
which changes my perspective on the inline ellipsis.

Perhaps the ellipsis regex should be:
```
r"^\s*(def .*: )?\.\.\.\s*(#|$)",
```

**Comment by KSmanis** [comment]
Off the top of my head:

Some false negatives:
```python
class AsyncProto(Protocol):
    async def __call__(self): ...


class MultilineProto(Protocol):
    def __call__(
        self,
        some_args,
        more_args,
    ): ...


class TypedProto(Protocol):
    def __call__(self) -> None: ...
```

A false positive:
```python
class Model(PydanticModel):
     long_field = Field(
         ...,
         description="long field",
     )
```

**Comment by nedbat** [comment]
The false negatives are great examples, thanks.  The false positive wouldn't match because of the comma, right?

**Comment by KSmanis** [comment]
True, the false positive could be only triggered by an unformatted (and quite odd):
```python
class Model(PydanticModel):
     field = Field(
         ...
     )
```

**Comment by cjw296** [comment]
@KSmanis - let's not let perfect be the enemy of good - your example is weird enough I'd be happy for the fix for it to be "please format your code differently" :-) 

(if I ended up with a need for that layout, I'd be totally fine with a custom exclude or just comment exclude!)

**Comment by KSmanis** [comment]
Of course, I'm not advocating for it to be considered. IMO a regex for Protocols and overloads would cover most common cases, rather than trying to perfect a regex for Ellipsis.

**Comment by cjw296** [comment]
Separate refex in the defaults to cover those two is a great suggestion!

**Comment by nedbat** [comment]
I'm still concerned that matching the words "overload" or "Protocol" will have false positives.  People can configure those extra regexes themselves.  Am I wrong to be concerned?

**Comment by nedbat** [comment]
In commit 4d4ddd87ddfa59c9d7b88c2c2d9fc6183668b483 I made the ellipsis pattern (much) more elaborate:
```python
r"^\s*(((async )?def .*?)?\)(\s*->.*?)?:\s*)?\.\.\.\s*(#|$)",
```
This works for all the real cases we've identified, but still has a false positive:
```python
def f(): print("(well): ... #2 false positive!")
```
I think it strikes a good balance.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.10.0](https://pypi.org/project/coverage/7.10.0).

**Comment by nedbat** [comment]
Oops, the `if TYPE_CHECKING:` exclusion was wrong, fixed in commit 1e2f41a756191898d4f1936fa01165d52ec9e624

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.10.1](https://pypi.org/project/coverage/7.10.1).
